src/main/java/ui/components/pickers/MilestonePickerState.java

```diff
@@ -30,78 +34,236 @@ public MilestonePickerState(List<PickerMilestone> milestones, String userInput)
-        return this.currentMilestonesList;
+    /**
+     * Generates the best matching milestones which will then be added to bestMatchingMilestones
+     *
+     * The added milestones are references to the actual elements in allMilestones, and thus should not be
+     * unnecessarily mutated
+     * @param userInput
+     */
+    private void determineBestMatchingMilestones(String userInput) {
+        bestMatchingMilestones.clear();
+        addMatchingMilestones(bestMatchingMilestones, getRemainingBestMatchesToLimit(), allMilestones);
+        if (isBelowLimit(bestMatchingMilestones, BEST_MATCHING_LIMIT)) {
+            addPreviouslyMatchingMilestones(bestMatchingMilestones, getRemainingBestMatchesToLimit(), userInput,
+                                            allMilestones);
+        }
+        if (isBelowLimit(bestMatchingMilestones, BEST_MATCHING_LIMIT)) {
+            addLikelyUnmatchedMilestones(bestMatchingMilestones, getRemainingBestMatchesToLimit(), allMilestones);
+        }
+    }
+
+    private boolean isBelowLimit(List<PickerMilestone> milestones, int limit) {
+        return milestones.size() < limit;
+    }
+
+    private int getRemainingBestMatchesToLimit() {
+        return BEST_MATCHING_LIMIT - bestMatchingMilestones.size();
+    }
+
+    /**
+     * Adds up to {@code limit} number of milestones to {@code bestMatchingMilestones}
+     * These milestones are taken from {@code milestones}, in the same order given,
+     * if they are not already in {@code bestMatchingMilestones}
+     *
+     * This method will do nothing if {@code limit} is not a positive integer
+     * @param bestMatchingMilestones
+     * @param limit
+     * @param milestones
+     */
+    private void addLikelyUnmatchedMilestones(List<PickerMilestone> bestMatchingMilestones, int limit,
+                                              List<PickerMilestone> milestones) {
+        if (limit <= 0) return;
+        List<PickerMilestone> likelyUnmatchedMilestones =
+                getLimitedFirstMilestones(bestMatchingMilestones, limit, milestones);
+        bestMatchingMilestones.addAll(likelyUnmatchedMilestones);
     }
 
     /**
-     * Gets the list of milestones that matches the current query
+     * Adds up to {@code limit} number of milestones to {@code bestMatchingMilestones}
+     * These milestones are taken from {@code milestones} if they match subsets of {@code userInput}, and are
+     * not already in {@code bestMatchingMilestones}
      *
-     * @return
+     * This method will do nothing if {@code limit} is not a positive integer
+     * @param bestMatchingMilestones
+     * @param limit
+     * @param userInput
+     * @param milestones
      */
-    public List<PickerMilestone> getMatchingMilestonesList() {
-        return this.currentMilestonesList.stream()
-                .filter(milestone -> !milestone.isMatching())
+    private void addPreviouslyMatchingMilestones(List<PickerMilestone> bestMatchingMilestones, int limit,
```

Comment:
Method name here is quite confusing. I thought it means from previous query, while it actually means user input that excludes last few characters. Maybe `LikelyMatchedMilestones` as opposite to `LikelyUnmatchedMilestones`?